### DISCUSSION TITLE
SPI Slave Sketch Example

### DISCUSSION THREAD

**Discussion by nbelles** [SpenceKonde/megaTinyCore#901] [Q&A] [answered]
Hello,

I have not used the SPI interface much at all and definitely not as a slave and am looking for a simple example writing data from the master to the slave. I have found some examples for generic Arduino's around the internet but I'm not sure that they work with this core. Any help would be greatly appreciated.

Thanks!

**Comment by SpenceKonde** [comment, marked as answer]
Spi slave on arduino is hard to pull off within the arduino api. It must be
implemented dmanually (look around, nobody else has an spi slave library
either. You've got to do it manually (for example, attached interrupts are
dead before the arrived, because if you are calling a function that it
can't inline, from the isr, the interrupt overhead is such that the timing
constraints at typical spi clock speed simply cannot be met. Even with
tight hand written isrs, it's tough. The core of the problem is thst you
can't clock stretch like i2c and to compound thst issue, spi clock speeds
are often higher. Even with buffered mode and manual interrup, if the sck
clock isn't on the low side, it's just not going to work because the
arduino can't get data into the buffer before the master tries to clock in
or out the next byte. I recommend wire slave, usart, or basically any other
means of communication unless you are a real wizard, and/or are flagrantly
abusing the spi module (as in, you're slave, but you're just using it as a
shift register and being your own master, combing with evsys and ccl to to
output those weirdo custom.protocols.

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by nbelles** [reply]
Sounds good, I'll stick to I2C then. Thanks for the info!

**Reply by SpenceKonde** [reply]
That's a wise decision. 

SPI slave mode is something that should be used only in extreme applications. 

I2C for master-slave communications in normal applications, and UART for peer to peer communications under normal and most abnormal applications.
